File: src/lib/protection.ts
import { t } from "@/lib/translations";

/**
 * Crop protection records: what was put on a field to deal with a pest, when,
 * and by which of the two approaches.
 *
 * The type and its sanitiser live here rather than in `field-data.ts` for the
 * same reason the soil and irrigation ones do — the module owns its own shape
 * and the shared document just composes them. Nothing here touches `node:fs` or
 * Leaflet, so the API route, the store and the browser all validate against
 * this file.
 */

export const PROTECTION_METHODS = ["biological", "chemical"] as const;
export type ProtectionMethod = (typeof PROTECTION_METHODS)[number];

export interface ProtectionLog {
  id: string;
  fieldId: string;
  // Denormalised so a record still reads correctly after its field is renamed
  // or deleted; the id is what links it back while the field still exists.
  fieldName: string;
  /** The day of the treatment, as `YYYY-MM-DD`. */
  date: string;
  method: ProtectionMethod;
  /**
   * What was applied, in the grower's own words — a product name for a chemical
   * treatment, the released beneficial for a biological one.
   *
   * It is one field rather than two because it answers one question, and the
   * method already says how to read it. Two mutually exclusive columns would
   * mean every record carrying a blank one, and a form that hides half of
   * itself. Free text is deliberate: pesticide registrations differ by country
   * and change every season, and a list compiled here would be wrong somewhere
   * from the day it was written.
   */
  agent: string;
  /** The pest or disease being treated. */
  target: string;
  /**
   * Free text as well, because the units are not comparable: a pesticide is
   * dosed in ml or g per decare, a beneficial is released in individuals or
   * cards per decare. Storing a number would mean picking one and mangling the
   * other.
   */
  dose: string;
  notes: string;
}

const MAX_TEXT = 200;
const MAX_NAME = 128;

function readString(value: unknown, max: number): string {
  return typeof value === "string" ? value.slice(0, max) : "";
}

/** Keeps plain calendar days only; anything else becomes "no date". */
function isoDay(value: unknown): string {
  if (typeof value !== "string") return "";
  const day = value.slice(0, 10);
  if (!/^\d{4}-\d{2}-\d{2}$/.test(day)) return "";
  return Number.isNaN(Date.parse(day)) ? "" : day;
}

/**
 * An unreadable method falls back to chemical. There is no neutral third state
 * to fall back to, and of the two that is both the commoner treatment and the
 * one whose record matters more — a residue question is asked of sprays, not of
 * released predators.
 */
function normalizeMethod(value: unknown): ProtectionMethod {
  return value === "biological" ? "biological" : "chemical";
}

export function sanitizeProtectionLogs(value: unknown): ProtectionLog[] {
  if (!Array.isArray(value)) return [];
  const result: ProtectionLog[] = [];
  for (const entry of value) {
    if (!entry || typeof entry !== "object") continue;
    const source = entry as Record<string, unknown>;
    if (typeof source.id !== "string" || source.id === "") continue;
    result.push({
      id: source.id.slice(0, 64),
      fieldId: readString(source.fieldId, 64),
      fieldName: readString(source.fieldName, MAX_NAME),
      date: isoDay(source.date),
      method: normalizeMethod(source.method),
      agent: readString(source.agent, MAX_NAME),
      target: readString(source.target, MAX_NAME),
      dose: readString(source.dose, 64),
      notes: readString(source.notes, MAX_TEXT),
    });
  }
  return result;
}

export function methodLabel(method: ProtectionMethod): string {
  return t.protection.methods[method];
}

/** What the "what was applied" field is called, which is method-dependent. */
export function agentLabel(method: ProtectionMethod): string {
  return method === "biological" ? t.protection.agentBiological : t.protection.agentChemical;
}

export function agentPlaceholder(method: ProtectionMethod): string {
  return method === "biological"
    ? t.protection.agentBiologicalPlaceholder
    : t.protection.agentChemicalPlaceholder;
}

export function dosePlaceholder(method: ProtectionMethod): string {
  return method === "biological"
    ? t.protection.doseBiologicalPlaceholder
    : t.protection.doseChemicalPlaceholder;
}

export function methodBadgeClass(method: ProtectionMethod): string {
  return method === "biological"
    ? "bg-emerald-100 text-emerald-700 dark:bg-emerald-950/60 dark:text-emerald-300"
    : "bg-amber-100 text-amber-700 dark:bg-amber-950/60 dark:text-amber-300";
}

/** Newest first; records without a usable date sort to the bottom. */
export function byNewestTreatment(a: ProtectionLog, b: ProtectionLog): number {
  const left = Date.parse(a.date);
  const right = Date.parse(b.date);
  if (Number.isNaN(left) && Number.isNaN(right)) return 0;
  if (Number.isNaN(left)) return 1;
  if (Number.isNaN(right)) return -1;
  return right - left;
}

/**
 * Every treatment recorded for one field, newest first. A blank id matches
 * nothing on purpose: it is what an empty field list leaves in a form, not a
 * field whose history should be shown.
 */
export function treatmentsForField(logs: ProtectionLog[], fieldId: string): ProtectionLog[] {
  if (fieldId === "") return [];
  return logs.filter((log) => log.fieldId === fieldId).sort(byNewestTreatment);
}

/** The form state. */
export type ProtectionDraft = Omit<ProtectionLog, "id" | "fieldName">;

export function createDraft(fieldId: string): ProtectionDraft {
  return {
    fieldId,
    date: new Date().toISOString().slice(0, 10),
    method: "chemical",
    agent: "",
    target: "",
    dose: "",
    notes: "",
  };
}

File: backend/crates/domain/src/protection.rs
//! Crop protection records — the Rust half of `src/lib/protection.ts`.
//!
//! What was put on a field to deal with a pest, when, and by which of the two
//! approaches. The split between what has been carried out and what is only
//! planned is the part the assistant turns on: advising on the strength of a
//! spray that was only ever pencilled in is the failure worth designing
//! against, so [`split_by_status`] exists here as well as in the browser.

use std::cmp::Ordering;

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::irrigation::iso_day;
use crate::js;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ProtectionMethod {
    Biological,
    Chemical,
}

impl ProtectionMethod {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Biological => "biological",
            Self::Chemical => "chemical",
        }
    }

    pub fn parse(value: &str) -> Option<Self> {
        match value {
            "biological" => Some(Self::Biological),
            "chemical" => Some(Self::Chemical),
            _ => None,
        }
    }

    /// Spelled out rather than left as a key, because it is the thing an
    /// answer turns on: a named organism means something different released as
    /// a beneficial than it would as a product name.
    pub fn label(self) -> &'static str {
        match self {
            Self::Biological => "Biological control",
            Self::Chemical => "Chemical control",
        }
    }

    /// An unreadable method falls back to chemical. There is no neutral third
    /// state, and of the two that is both the commoner treatment and the one
    /// whose record matters more — a residue question is asked of sprays, not
    /// of released predators.
    fn normalize(value: Option<&Value>) -> Self {
        if js::as_str(value) == Some("biological") { Self::Biological } else { Self::Chemical }
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ProtectionStatus {
    Planned,
    Applied,
}

impl ProtectionStatus {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Planned => "planned",
            Self::Applied => "applied",
        }
    }

    /// An unreadable status falls back to applied, for two reasons that point
    /// the same way. Records written before the module told the two apart were
    /// all treatments already carried out. And of the two mistakes, reading a
    /// real application as merely planned is the dangerous one: it invites a
    /// second spray of something already sprayed.
    fn normalize(value: Option<&Value>) -> Self {
        if js::as_str(value) == Some("planned") { Self::Planned } else { Self::Applied }
    }
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProtectionLog {
    pub id: String,
    pub field_id: String,
    // Denormalised so a record still reads correctly after its field is renamed
    // or deleted; the id is what links it back while the field still exists.
    pub field_name: String,
    /// Whether the treatment has happened or is still due.
    pub status: ProtectionStatus,
    /// The day of the treatment, as `YYYY-MM-DD` — the day it was carried out,
    /// or the day it is planned for. Which one it means is what `status` says.
    pub date: String,
    pub method: ProtectionMethod,
    /// What was applied, in the grower's own words.
    pub agent: String,
    /// The pest or disease being treated.
    pub target: String,
    /// Free text, because the units are not comparable: a pesticide is dosed in
    /// ml or g per decare, a beneficial in individuals or cards per decare.
    pub dose: String,
    pub notes: String,
}

const MAX_TEXT: usize = 200;
const MAX_NAME: usize = 128;

pub fn sanitize_protection_logs(value: Option<&Value>) -> Vec<ProtectionLog> {
    let Some(Value::Array(entries)) = value else {
        return Vec::new();
    };
    let mut result = Vec::new();
    for entry in entries {
        let Some(source) = entry.as_object() else {
            continue;
        };
        let id = match js::as_str(source.get("id")) {
            Some(id) if !id.is_empty() => js::slice_utf16(id, 64),
            _ => continue,
        };
        result.push(ProtectionLog {
            id,
            field_id: js::text(source.get("fieldId"), 64),
            field_name: js::text(source.get("fieldName"), MAX_NAME),
            status: ProtectionStatus::normalize(source.get("status")),
            date: iso_day(source.get("date")),
            method: ProtectionMethod::normalize(source.get("method")),
            agent: js::text(source.get("agent"), MAX_NAME),
            target: js::text(source.get("target"), MAX_NAME),
            dose: js::text(source.get("dose"), 64),
            notes: js::text(source.get("notes"), MAX_TEXT),
        });
    }
    result
}

/// Newest first; records without a usable date sort to the bottom.
pub fn by_newest_treatment(left: &ProtectionLog, right: &ProtectionLog) -> Ordering {
    js::newest_first(js::date_parse(&left.date), js::date_parse(&right.date))
}

/// Soonest first; records without a usable date sort to the bottom.
pub fn by_soonest_planned(left: &ProtectionLog, right: &ProtectionLog) -> Ordering {
    js::soonest_first(js::date_parse(&left.date), js::date_parse(&right.date))
}

/// Every treatment recorded for one field, newest first. A blank id matches
/// nothing on purpose: it is what an empty field list leaves in a form, not a
/// field whose history should be shown.
pub fn treatments_for_field<'a>(
    logs: &'a [ProtectionLog],
    field_id: &str,
) -> Vec<&'a ProtectionLog> {
    if field_id.is_empty() {
        return Vec::new();
    }
    let mut found: Vec<&ProtectionLog> =
        logs.iter().filter(|log| log.field_id == field_id).collect();
    found.sort_by(|left, right| by_newest_treatment(left, right));
    found
}

pub struct ProtectionSplit<'a> {
    pub planned: Vec<&'a ProtectionLog>,
    pub applied: Vec<&'a ProtectionLog>,
}

/// The two states, each in the order that state is read in.
///
/// A plan is about the future, so the nearest one matters most and they run
/// soonest first; a record of what was done is about the past, so the most
/// recent matters most and they run newest first. One mixed list ordered
/// either way buries half of itself.
pub fn split_by_status<'a, I>(logs: I) -> ProtectionSplit<'a>
where
    I: IntoIterator<Item = &'a ProtectionLog>,
{
    let mut planned = Vec::new();
    let mut applied = Vec::new();
    for log in logs {
        match log.status {
            ProtectionStatus::Planned => planned.push(log),
            ProtectionStatus::Applied => applied.push(log),
        }
    }
    planned.sort_by(|left, right| by_soonest_planned(left, right));
    applied.sort_by(|left, right| by_newest_treatment(left, right));
    ProtectionSplit { planned, applied }
}
